cpp/src/strings/split/split.cuh

```diff
@@ -0,0 +1,407 @@
+    cudf::detail::get_value<size_type>(input.offsets(), input.offset(), stream);
+
+  auto d_offsets = input.offsets_begin();
+
+  // count the number of delimiters in the entire column
+  auto const delimiter_count =
+    thrust::count_if(rmm::exec_policy(stream),
+                     thrust::make_counting_iterator<size_type>(0),
+                     thrust::make_counting_iterator<size_type>(chars_bytes),
+                     [tokenizer, d_offsets, chars_bytes] __device__(size_type idx) {
+                       return tokenizer.is_delimiter(idx, d_offsets, chars_bytes);
+                     });
+  // Create a vector of every delimiter position in the chars column.
+  // These may include overlapping or otherwise out-of-bounds delimiters which
+  // will be resolved during token processing.
+  auto delimiter_positions = rmm::device_uvector<size_type>(delimiter_count, stream);
+  auto d_positions         = delimiter_positions.data();
+  auto const copy_end =
+    thrust::copy_if(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator<size_type>(0),
+                    thrust::make_counting_iterator<size_type>(chars_bytes),
+                    delimiter_positions.begin(),
+                    [tokenizer, d_offsets, chars_bytes] __device__(size_type idx) {
+                      return tokenizer.is_delimiter(idx, d_offsets, chars_bytes);
+                    });
+
+  // create a vector of offsets to each string's delimiter set within delimiter_positions
+  auto const delimiter_offsets = [&] {
+    // first, create a vector of string indices for each delimiter
+    auto string_indices = rmm::device_uvector<size_type>(delimiter_count, stream);
+    thrust::upper_bound(rmm::exec_policy(stream),
+                        d_offsets,
+                        d_offsets + strings_count,
+                        delimiter_positions.begin(),
+                        copy_end,
+                        string_indices.begin());
+
+    // compute delimiter offsets per string
+    auto delimiter_offsets   = rmm::device_uvector<size_type>(strings_count + 1, stream);
+    auto d_delimiter_offsets = delimiter_offsets.data();
+
+    // memset to zero-out the delimiter counts for any null-entries or strings with no delimiters
+    CUDF_CUDA_TRY(cudaMemsetAsync(
+      d_delimiter_offsets, 0, delimiter_offsets.size() * sizeof(size_type), stream.value()));
+
+    // next, count the number of delimiters per string
+    auto d_string_indices = string_indices.data();  // identifies strings with delimiters only
+    thrust::for_each_n(rmm::exec_policy(stream),
+                       thrust::make_counting_iterator<size_type>(0),
+                       delimiter_count,
+                       [d_string_indices, d_delimiter_offsets] __device__(size_type idx) {
+                         auto const str_idx = d_string_indices[idx] - 1;
+                         atomicAdd(d_delimiter_offsets + str_idx, 1);
+                       });
```

Comment:
Could this just be a `thrust::reduce_by_key`? If I'm reading this right, the inputs are a string index per delimiter, ie:

`0, 0, 1, 1, 1, 3, 3, 3, 5, 5, 6`

A `reduce_by_key` with the keys being the above and the value just a constant_iterator returning 1 should yield the same result without the atomics.

Reply:
Although I guess you'd have to deal with the holes for strings that have no delimiters.

Reply:
Right. Some of the strings have no delimiters and we need zero for those.